types/parser_driver/value_expr.go

```diff
@@ -69,11 +68,39 @@ type ValueExpr struct {
 	projectionOffset int
 }
 
-// Restore implements Recoverable interface.
-func (n *ValueExpr) Restore(sb *strings.Builder) error {
-	err := n.format(sb)
-	if err != nil {
-		return errors.Trace(err)
+// Restore implements Node interface.
+func (n *ValueExpr) Restore(ctx *ast.RestoreCtx) error {
+	switch n.Kind() {
+	case types.KindNull:
+		ctx.WriteKeyWord("NULL")
+	case types.KindInt64:
+		if n.Type.Flag&mysql.IsBooleanFlag != 0 {
+			if n.GetInt64() > 0 {
+				ctx.WriteKeyWord("TRUE")
+			} else {
+				ctx.WriteKeyWord("FALSE")
+			}
+		} else {
+			ctx.WritePlain(strconv.FormatInt(n.GetInt64(), 10))
+		}
+	case types.KindUint64:
+		ctx.WritePlain(strconv.FormatUint(n.GetUint64(), 10))
+	case types.KindFloat32:
+		ctx.WritePlain(strconv.FormatFloat(n.GetFloat64(), 'e', -1, 32))
+	case types.KindFloat64:
+		ctx.WritePlain(strconv.FormatFloat(n.GetFloat64(), 'e', -1, 64))
+	case types.KindString, types.KindBytes:
+		ctx.WriteString(n.GetString())
+	case types.KindMysqlDecimal:
+		ctx.WritePlain(n.GetMysqlDecimal().String())
+	case types.KindBinaryLiteral:
+		if n.Type.Flag&mysql.UnsignedFlag != 0 {
+			_, _ = fmt.Fprintf(ctx.In, "x'%x'", n.GetBytes())
```

Comment:
IMO it is better to 

```go
ctx.WritePlain(fmt.Sprintf("x'%x'", n.GetBytes()))
```

or expose a new function

```go
ctx.WriteFormat("x'%x'", n.GetBytes())
```

than directly using the field `ctx.In`.